### PR TITLE
PS-7806: Column compression breaks async replication on PS

### PR DESCRIPTION
**Pull request by nitendra-bhosle**
https://jira.percona.com/browse/PS-7806

Analysis
compress_heap memory was released inside
method row_sel_store_mysql_rec , even though ,it was reused.
Which caused incorrect bin logging for comressed column,
leading to async replication to break.

Fix

Removed compress_heap free statements,
hence memory is reused.

### CODE REVIEW

**Comment by nitendra-bhosle** [comment]
Jenkins : https://ps80.cd.percona.com/view/8.0/job/percona-server-8.0-param/1050/

File: storage/innobase/row/row0sel.cc
Comment (satya-bodapati) [inline storage/innobase/row/row0sel.cc:2991]: Can you plz explain the following?
1. How it is reused? This function is supposed to transfer/convert a innodb record to server/mysql record
2. When is it freed?
3. How about emptying it? Like blob_heap above. Can't we do the same?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: mysql-test/suite/binlog/t/binlog_stm_delete_with_compress_column.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/binlog/t/binlog_stm_delete_with_compress_column.test]: I don't think there is no any need to run the test in statement format, as corrupted data is not propagated in statement based replication. So, I suggest you to please remove this test.

File: mysql-test/extra/binlog_tests/binlog_delete_with_compress_column.test
Comment (venkatesh-prasad-v) [inline mysql-test/extra/binlog_tests/binlog_delete_with_compress_column.test]: The tests pass even without fix, so please add a replication test (rpl suite) as this bug was initially found in a replication scenario.
```
------------------------------------------------------------------------------
[ 14%] binlog.binlog_row_delete_with_compress_column 'mix'  [ skipped ]  Doesn't support --binlog-format = 'mixed'
[ 28%] binlog.binlog_stm_delete_with_compress_column 'mix'  [ pass ]   4712
[ 42%] binlog.binlog_row_delete_with_compress_column 'row'  [ pass ]   5768
[ 57%] binlog.binlog_stm_delete_with_compress_column 'row'  [ skipped ]  Doesn't support --binlog-format = 'row'
[ 71%] binlog.binlog_row_delete_with_compress_column 'stmt'  [ skipped ]  Doesn't support --binlog-format = 'statement'
[ 85%] binlog.binlog_stm_delete_with_compress_column 'stmt'  [ pass ]   4456
[100%] shutdown_report                           [ pass ]
```

**Review by venkatesh-prasad-v — CHANGES_REQUESTED** [review]

File: storage/innobase/row/row0sel.cc
Comment (satya-bodapati) [inline storage/innobase/row/row0sel.cc:2991]: If we are reusing, even emptying is a problem isn't it? I think partitions are supposed to their partition-specific blob heaps? Can you please check again?

**Review by satya-bodapati — APPROVED** [review]

File: storage/innobase/row/row0sel.cc
Comment (satya-bodapati) [reply storage/innobase/row/row0sel.cc:2991]: Sorry used the wrong button.. "not approved yet" :)

File: mysql-test/suite/rpl/t/rpl_column_compression_async_rpl.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/rpl/t/rpl_column_compression_async_rpl.test]: You can remove these lines. The rationale is 
1. `--source include/master-slave.inc` by default leaves the connection on master, so there is no need to change the connection to master.
2. As we haven't done anything in the test, there is nothing to sync.

File: mysql-test/suite/rpl/t/rpl_column_compression_async_rpl.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/rpl/t/rpl_column_compression_async_rpl.test:5]: Also, please source `include/have_binlog_format_row.inc` as there is no need to run this test in all formats.

File: mysql-test/suite/rpl/t/rpl_column_compression_async_rpl.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/rpl/t/rpl_column_compression_async_rpl.test]: This is not required.

File: mysql-test/suite/rpl/t/rpl_column_compression_async_rpl.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/rpl/t/rpl_column_compression_async_rpl.test]: Please source `sync_slave_sql_with_master.inc` file to make sure that the updates reach replica and are applied correctly on replica.

File: mysql-test/suite/rpl/t/rpl_column_compression_async_rpl.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/rpl/t/rpl_column_compression_async_rpl.test]: Can you please use `rpl_diff.inc` to ensure data consistency across both the source and replica.

**Review by venkatesh-prasad-v — CHANGES_REQUESTED** [review]

File: mysql-test/suite/rpl/t/rpl_column_compression_async_rpl.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/rpl/t/rpl_column_compression_async_rpl.test:4]: Please change the order of inclusion of inc files.
```suggestion
--source include/have_binlog_format_row.inc
--source include/master-slave.inc
```

File: mysql-test/suite/rpl/t/rpl_column_compression_async_rpl.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/rpl/t/rpl_column_compression_async_rpl.test]: Please source `rpl_connection_master.inc` instead.

**Review by venkatesh-prasad-v — CHANGES_REQUESTED** [review]

**Review by venkatesh-prasad-v — APPROVED** [review]
Looks fine with the test. But please wait for final approval from @satya-bodapati

File: storage/innobase/row/row0sel.cc
Comment (satya-bodapati) [reply storage/innobase/row/row0sel.cc:2991]: @nitendra-bhosle can you please reply to these questions?

I don't understand the issue. Specifically why free is a problem. and empty is not?
AFAIU, empty doesn't the remove the first block. So may be the compressed heap is staying within the first block of heap.

Can you please create a blob text such that it occupies more than the first block in heap?

This is outdated changes.

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
Please see the question on thread

**Comment by nitendra-bhosle** [comment]
@satya-bodapati  :  https://github.com/percona/percona-server/pull/4460#discussion_r714653808 
This is outdated. I have removed both empty and free heap statements.

**Comment by satya-bodapati** [comment]
> @satya-bodapati : [#4460 (comment)](https://github.com/percona/percona-server/pull/4460#discussion_r714653808)
> This is outdated. I have removed both empty and free heap statements.

Well, I still the empty function.

If you remove this empty, Do we call empty at the end of statement execution?
Can you explain how the 'reuse' is happening?

**Comment by nitendra-bhosle** [comment]
> > @satya-bodapati : [#4460 (comment)](https://github.com/percona/percona-server/pull/4460#discussion_r714653808)
> > This is outdated. I have removed both empty and free heap statements.
> 
> Well, I still the empty function.
> 
> If you remove this empty, Do we call empty at the end of statement execution?

This empty statement , I have added just to be consistent with blob_heap .
If we remove this, we will not call empty at the end of statement execution, as it's already being 
getting free's inside : row_prebuilt_free 

> Can you explain how the 'reuse' is happening?

We are allocating memory column_get_compress_header and row_compress_column 
using statements like : 
if (!prebuilt->compress_heap)
    prebuilt->compress_heap = mem_heap_create(ut_max(UNIV_PAGE_SIZE, buf_len));

I reffered reuse , because this is diverging from blob_heap memory management in a sense that memory is getting
allowcated inside these two methods , and reuse might be happening , otherwise , no need to allocated and check whether it's already allocated inside other method.

**Comment by bratao** [comment]
Any chance of this making into 8.0.26 release? I just got bitten by this

**Comment by nitendra-bhosle** [comment]
> Any chance of this making into 8.0.26 release? I just got bitten by this

I think this Issue will be part of 8.0.27 release

File: storage/innobase/row/row0sel.cc
Comment (satya-bodapati) [reply storage/innobase/row/row0sel.cc:2991]: Can you please bring this back and test? Only then we will know if your new fix is working or not?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
Also please test PS-7657 by reverting the fix and using this fix

**Comment by nitendra-bhosle** [comment]
> Also please test PS-7657 by reverting the fix and using this fix

Tested , All tests are passing

**Comment by satya-bodapati** [comment]
> > Also please test PS-7657 by reverting the fix and using this fix
> 
> Tested , All tests are passing

Thats great news! Please include the revert of PS-7657 code change in this PR

**Comment by nitendra-bhosle** [comment]
> > > Also please test PS-7657 by reverting the fix and using this fix
> > 
> > 
> > Tested , All tests are passing
> 
> Thats great news! Please include the revert of PS-7657 code change in this PR

Revert is conflicting with this change : 
if (UNIV_LIKELY_NULL(prebuilt->compress_heap))
    row_mysql_prebuilt_free_compress_heap(prebuilt);

Earlier we had 
if (UNIV_LIKELY_NULL(prebuilt->compress_heap))
mem_heap_empty(prebuilt->compress_heap);

**Comment by satya-bodapati** [comment]
> > > > Also please test PS-7657 by reverting the fix and using this fix
> > > 
> > > 
> > > Tested , All tests are passing
> > 
> > 
> > Thats great news! Please include the revert of PS-7657 code change in this PR
> 
> Revert is conflicting with this change : if (UNIV_LIKELY_NULL(prebuilt->compress_heap)) row_mysql_prebuilt_free_compress_heap(prebuilt);
> 
> Earlier we had if (UNIV_LIKELY_NULL(prebuilt->compress_heap)) mem_heap_empty(prebuilt->compress_heap);

You can add a new commit which will bring to original state.

**Comment by nitendra-bhosle** [comment]
Also this change of PS-7657 is still required - 
if (UNIV_LIKELY_NULL(prebuilt->compress_heap))
    mem_heap_empty(prebuilt->compress_heap);

Inside function row_insert_for_mysql_using_ins_graph

**Comment by satya-bodapati** [comment]
> Also this change of PS-7657 is still required - if (UNIV_LIKELY_NULL(prebuilt->compress_heap)) mem_heap_empty(prebuilt->compress_heap);
> 
> Inside function row_insert_for_mysql_using_ins_graph

Yes @nitendra-bhosle, please bring that emptying back. with the new logic, emptying or freeing compress_heap shouldn't matter.

**Comment by satya-bodapati** [comment]
> > Also this change of PS-7657 is still required - if (UNIV_LIKELY_NULL(prebuilt->compress_heap)) mem_heap_empty(prebuilt->compress_heap);
> > Inside function row_insert_for_mysql_using_ins_graph
> 
> Yes @nitendra-bhosle, please bring that emptying back. with the new logic, emptying or freeing compress_heap shouldn't matter.

@nitendra-bhosle I still don't see the emptying code in row_insert_for_mysql_using_ins_graph()
```
  if (UNIV_LIKELY_NULL(prebuilt->compress_heap))
    mem_heap_empty(prebuilt->compress_heap);
```
This was present before PS-7657 and it should be brought back. It will not have any effect with your latest fix.

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
Please check my latest comment

**Comment by nitendra-bhosle** [comment]
> > Also this change of PS-7657 is still required - if (UNIV_LIKELY_NULL(prebuilt->compress_heap)) mem_heap_empty(prebuilt->compress_heap);
> > Inside function row_insert_for_mysql_using_ins_graph
> 
> Yes @nitendra-bhosle, please bring that emptying back. with the new logic, emptying or freeing compress_heap shouldn't matter.




> > > Also this change of PS-7657 is still required - if (UNIV_LIKELY_NULL(prebuilt->compress_heap)) mem_heap_empty(prebuilt->compress_heap);
> > > Inside function row_insert_for_mysql_using_ins_graph
> > 
> > 
> > Yes @nitendra-bhosle, please bring that emptying back. with the new logic, emptying or freeing compress_heap shouldn't matter.
> 
> @nitendra-bhosle I still don't see the emptying code in row_insert_for_mysql_using_ins_graph()
> 
> ```
>   if (UNIV_LIKELY_NULL(prebuilt->compress_heap))
>     mem_heap_empty(prebuilt->compress_heap);
> ```
> 
>This was present before PS-7657 and it should be brought back. It will not have any effect with your latest fix.

After my investigation today, I found that above change is required to fix - PS-7657.It will not have any effect on my fix for PS-7806.
But this will break for PS-7657 , hence I kept it as is.

With current patch , which is part of these commits , all tests for PS--7657 and PS-7806 are passing

File: storage/innobase/row/row0mysql.cc
Comment (satya-bodapati) [inline storage/innobase/row/row0mysql.cc:2035]: previously before 7657, the empty was before trx->op_info = "inserting" line. Any reason, it is placed here? Does it have to be done after conversion of row to innodb format?

Can it be moved to before trx->op_info?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

**Comment by nitendra-bhosle** [comment]
> previously before 7657, the empty was before trx->op_info = "inserting" line. Any reason, it is placed here? Does it have to be done after conversion of row to innodb format?
> 
> Can it be moved to before trx->op_info?

We can't move it before trx->op_info , as row_mysql_convert_row_to_innobase call uses prebuilt->compress_heap

File: storage/innobase/row/row0mysql.cc
Comment (satya-bodapati) [reply storage/innobase/row/row0mysql.cc:2035]: I see that you replied below.
So this means, we cannot free a buffer from the previous handler call and it is still in use?

Ideally, we should be able to free **before** the mysql_to_innodb row format conversion. The buffer would have last row insertion/handler call. So technically we should be allowed to free it.

I think I asked this question (may be on slack), can you please add empty or freeing of normal heap before trx->op_info/row_mysql_convert_row_to_innobase()? Will it have the same problem if we free earlier?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
Please check my last comment

**Comment by kamil-holubicki** [comment]
Will be handled by 
https://github.com/percona/percona-server/pull/5161
https://github.com/percona/percona-server/pull/5162

**Pull request closed without merging**
